blockchain/reactor_test.go

```diff
@@ -0,0 +1,218 @@
+	logBuf := new(bytes.Buffer)
+	logger := log.NewTMLogger(logBuf)
+	maxBlockHeight := 20
+
+	bcr := newBlockchainReactor(logger, maxBlockHeight)
+	go bcr.OnStart()
+	defer bcr.Stop()
+
+	// Add some peers in
+	peer := newbcrTestPeer(cmn.RandStr(12))
+	bcr.AddPeer(peer)
+
+	chID := byte(0x01)
+	
+	tests := [...]struct {
+		height   int
+		existent bool
+	}{
+		0: {
+			height:   maxBlockHeight + 2,
+			existent: false,
+		},
+		1: {
+			height:   10,
+			existent: true, // We should have this
+		},
+		2: {
+			height:   1,
+			existent: true, // We should have this
+		},
+		3: {
+			height:   100,
+			existent: false,
+		},
+	}
+
+	// Currently the repsonses take asynchronous paths so
+	// the results are skewed. However, we can ensure that
+	// the (expectedHeight, expectedExistence) all tally up.
+	heightTally := map[int]bool{}
+	var results []BlockchainMessage
+
+	msgRequest := byte(0x10)
+	for _, tt := range tests {
+		reqBlockBytes := []byte{msgRequest, 0x02, 0x00, byte(tt.height)}
+		bcr.Receive(chID, peer, reqBlockBytes)
+		heightTally[tt.height] = tt.existent
+		kv := peer.lastKeyValue()
+		results = append(results, kv.value.(struct{ BlockchainMessage }).BlockchainMessage)
+	}
+
+	for i, res := range results {
+		if bcRM, ok := res.(*bcBlockResponseMessage); ok {
+			block := bcRM.Block
+			if block == nil {
+				t.Errorf("result: #%d: expecting a non-nil block", i)
+				continue
+			}
+			mustExist, foundHeight := heightTally[block.Height]
+			if !foundHeight {
+				t.Errorf("result: #%d, missing height %d", i, block.Height)
+				continue
+			}
+			if !mustExist {
+				t.Errorf("result: #%d mustExist", i)
+			} else {
+				delete(heightTally, block.Height)
+			}
+		} else if bcNBRSM, ok := res.(*bcNoBlockResponseMessage); ok {
+			mustExist, foundHeight := heightTally[bcNBRSM.Height]
+			if !foundHeight {
+				t.Errorf("result: #%d, missing height %d", i, bcNBRSM.Height)
+				continue
+			}
+			if mustExist {
+				t.Errorf("result: #%d mustNotExist", i)
+			} else {
+				// Passes, remove this height from the tally
+				delete(heightTally, bcNBRSM.Height)
+			}
+		} else {
+			t.Errorf("result: #%d is an unexpected type: %T; data: %#v", i, res, res)
+		}
+	}
+
+	if len(heightTally) > 0 {
+		t.Errorf("Untallied heights: %#v\n", heightTally)
+	}
+}
+
+func makeTxs(blockNumber int) (txs []types.Tx) {
+	for i := 0; i < 10; i++ {
+		txs = append(txs, types.Tx([]byte{byte(blockNumber)}))
+	}
+	return txs
+}
+
+const testPartSize = 65536
+
+func makeBlock(blockNumber int, state *sm.State) *types.Block {
+	prevHash := state.LastBlockID.Hash
+	prevParts := types.PartSetHeader{}
+	valHash := state.Validators.Hash()
+	prevBlockID := types.BlockID{prevHash, prevParts}
+	block, _ := types.MakeBlock(blockNumber, "test_chain", makeTxs(blockNumber),
+		new(types.Commit), prevBlockID, valHash, state.AppHash, testPartSize)
+	return block
+}
+
+// The Test peer
+type bcrTestPeer struct {
+	key string
+	ch  chan *keyValue
+	kvm map[byte]interface{}
```

Comment:
not seeing where it is being used...

Reply:
Right, thanks for this catch. It is a vestige from the previous round of review.